### PR TITLE
Cloudwatch: QueryEditor rewrite to React

### LINKED ISSUE #19880: CloudWatch - Convert QueryEditor to React

### PR DESCRIPTION
**Pull request by sunker**
Fixes #19880

### CODE REVIEW

**Comment by shavonn** [comment]
This PR shouldn't be merged until after https://github.com/grafana/grafana/issues/19772 is completed so that Stats is allowed custom values.

**Review by sunker — COMMENTED** [review]
LGTM! 

Don't think the go.mod and go.sum should be checked in. Can you remove those?

File: public/app/plugins/datasource/cloudwatch/components/Forms.tsx [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+  tooltip?: string;
+}
+
+export const FormField: FunctionComponent<Partial<Props>> = ({ label, inputEl, className, tooltip, ...inputProps }) => (
+  <>
+    <FormLabel width={8} className="query-keyword" tooltip={tooltip}>
+      {label}
+    </FormLabel>
+
+    {inputEl || <Input type="text" className={`gf-form-input`} {...inputProps} />}
```
Comment (torkelo) [inline public/app/plugins/datasource/cloudwatch/components/Forms.tsx]: Mot sure I like this inputEl and that FormField has props from HTMLInputElement but you can change the inputEl to anything. 

(I think this is a big problem with the FormField in @grafana/ui as well and we are going to change that). 

Think it's better to let FormField just take label & tooltip as props and let inputEl be handled via children prop instead.

File: public/app/plugins/datasource/cloudwatch/components/Forms.tsx [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+  </>
+);
+
+export const InlineFormField: FunctionComponent<Props> = ({ ...props }) => {
```
Comment (torkelo) [inline public/app/plugins/datasource/cloudwatch/components/Forms.tsx]: If possible these FormField and InlineFormField could maybe be named QueryField and InlineQueryField? So we do no confuse them with general form fields components defined in @grafana/ui

File: public/app/plugins/datasource/cloudwatch/components/Stats.tsx [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+const options: Array<SelectableValue<string>> = ['Average', 'Maximum', 'Minimum', 'Sum', 'SampleCount'].map(value => ({
+  label: value,
+  value: value,
+}));
+
+const removeText = '-- remove stat --';
+const removeOption: SelectableValue<string> = { label: removeText, value: removeText };
+
+export const Stats: SFC<Props> = ({ values, onChange, variableOptionGroup }) => (
```
Comment (torkelo) [inline public/app/plugins/datasource/cloudwatch/components/Stats.tsx]: SFC is deprecated (use FunctionalComponent or the short alias FC)

**Review by torkelo — APPROVED** [review]
Looks good. Is this used from the annotation editor as well? I think cloudwatch used a shared component between query editor and annotation query editor. 

Just some minor comments

File: public/app/plugins/datasource/cloudwatch/components/Dimensions.tsx [hunk 4]
```diff
@@ -0,0 +1,77 @@
+import React, { FunctionComponent, Fragment, useState, useEffect } from 'react';
+import isEqual from 'lodash/isEqual';
+import { SelectableValue } from '@grafana/data';
+import { SegmentAsync } from '@grafana/ui';
+
+export interface Props {
+  dimensions: { [key: string]: string | string[] };
+  onChange: (dimensions: { [key: string]: string }) => void;
+  loadValues: (key: string) => Promise<Array<SelectableValue<string>>>;
```
[the remark this answers is not in the document]
Comment (sunker) [reply public/app/plugins/datasource/cloudwatch/components/Dimensions.tsx]: 👍

File: public/app/plugins/datasource/cloudwatch/components/Forms.tsx (hunk 1 above)
Comment (sunker) [reply public/app/plugins/datasource/cloudwatch/components/Forms.tsx]: I agree!

File: public/app/plugins/datasource/cloudwatch/components/Forms.tsx (hunk 2 above)
Comment (sunker) [reply public/app/plugins/datasource/cloudwatch/components/Forms.tsx]: yes!

File: public/app/plugins/datasource/cloudwatch/components/Stats.tsx (hunk 3 above)
Comment (sunker) [reply public/app/plugins/datasource/cloudwatch/components/Stats.tsx]: 👍

**Comment by sunker** [comment]
> Looks good. Is this used from the annotation editor as well? I think cloudwatch used a shared component between query editor and annotation query editor.
> 
> Just some minor comments

Auch, I had totally missed that the annotations editor uses parts of the query editor. I temporarily added the old editor back, but obviously that is not sustainable solution for the future. The framework has no support for rendering react AnnotationsQueryEditors right know? If not, would it be possible to add support for that in the near future? If not, what do you suggest we do? Create a AnnotationsQueryEditors and render it in an angular wrapper from within the plugin?

**Comment by sunker** [comment]
This PR was merged into #20268, so changes should go into that branch instead.

**Pull request closed without merging**